lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+		UnflushedBytes:                  make(map[uuid.UUID]UnflushedStateSyncerBytes),
+		MempoolKeyValueMap:              make(map[string][]byte),
+		MempoolFlushKeySet:              make(map[string]bool),
+		StateSyncerMutex:                &sync.Mutex{},
+		SyncType:                        nodeSyncType,
+		BlocksyncCompleteEntriesFlushed: blocksyncCompleteEntriesFlushed,
+	}
+}
+
+// Reset resets the state change syncer by truncating the state change file and index file.
+func (stateChangeSyncer *StateChangeSyncer) Reset() {
+	err := stateChangeSyncer.StateChangeFile.Truncate(0)
+	if err != nil {
+		glog.Fatalf("Error truncating stateChangeFile: %v", err)
+	}
+
+	err = stateChangeSyncer.StateChangeIndexFile.Truncate(0)
+	if err != nil {
+		glog.Fatalf("Error truncating stateChangeIndexFile: %v", err)
+	}
+
+	stateChangeSyncer.StateChangeFileSize = 0
+
+	stateChangeSyncer.EntryCount = 0
+}
+
+// handleDbTransactionConnected is called when a badger db operation takes place.
+// This function checks to see if the operation effects a "core_state" index, and if so it encodes a StateChangeEntry
+// to be written to the state change file upon DB flush.
+// It also writes the offset of the entry in the file to a separate index file, such that a consumer can look up a
+// particular entry index in the state change file.
+func (stateChangeSyncer *StateChangeSyncer) _handleStateSyncerOperation(event *StateSyncerOperationEvent) {
+	// If we're in blocksync mode, we only want to flush entries once the sync is complete.
+	if !stateChangeSyncer.BlocksyncCompleteEntriesFlushed && stateChangeSyncer.SyncType == NodeSyncTypeBlockSync {
+		return
+	}
+
+	stateChangeEntry := event.StateChangeEntry
+
+	// Check to see if the index in question has a "core_state" annotation in its definition.
+	if !isCoreStateKey(stateChangeEntry.KeyBytes) {
+		return
+	}
+
+	// Make sure 2 operations aren't logged simultaneously.
+	stateChangeSyncer.StateSyncerMutex.Lock()
+	defer stateChangeSyncer.StateSyncerMutex.Unlock()
+
+	flushId := event.FlushId
+
+	// Crate a block sync flush ID if one doesn't already exist.
+	if event.FlushId == uuid.Nil && stateChangeSyncer.BlockSyncFlushId == uuid.Nil {
+		stateChangeSyncer.BlockSyncFlushId = uuid.New()
+	}
+
+	if event.IsMempoolTxn {
+		// Create a mempool flush ID if one doesn't already exist.
+		if event.FlushId == uuid.Nil && stateChangeSyncer.MempoolFlushId == uuid.Nil {
+			stateChangeSyncer.MempoolFlushId = uuid.New()
+		}
+
+		// If the event flush ID is nil, then we need to use the global mempool flush ID.
+		if event.FlushId == uuid.Nil {
+			flushId = stateChangeSyncer.MempoolFlushId
+		}
+
+		// The current state of the tracked mempool is stored in the MempoolKeyValueMap. If this entry is already in there
+		// then we don't need to re-write it to the state change file.
+		// Create key for op + key map
+		txKey := fmt.Sprintf("%v%v", event.StateChangeEntry.OperationType, string(event.StateChangeEntry.KeyBytes))
```

Comment:
I'd make this into a function and use that function everywhere rather than manually Sprintf'ing.